File: jogl-samples/src/tests/gl_300/Gl_300_fbo_multisample.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package tests.gl_300;

import com.jogamp.opengl.GL;
import static com.jogamp.opengl.GL.GL_ARRAY_BUFFER;
import static com.jogamp.opengl.GL.GL_FLOAT;
import static com.jogamp.opengl.GL2ES3.*;
import com.jogamp.opengl.GL3;
import com.jogamp.opengl.util.GLBuffers;
import com.jogamp.opengl.util.glsl.ShaderCode;
import com.jogamp.opengl.util.glsl.ShaderProgram;
import glm.glm;
import glm.mat._4.Mat4;
import glm.vec._2.Vec2;
import framework.BufferUtils;
import framework.Profile;
import framework.Semantic;
import framework.Test;
import java.io.IOException;
import java.nio.FloatBuffer;
import java.util.logging.Level;
import java.util.logging.Logger;
import jgli.Gl;
import jgli.Load;
import glm.vec._2.i.Vec2i;
import glm.vec._4.Vec4;
import glf.Vertex_v2fv2f;
import java.nio.IntBuffer;

/**
 *
 * @author gbarbieri
 */
public class Gl_300_fbo_multisample extends Test {

    public static void main(String[] args) {
        Gl_300_fbo_multisample gl_300_fbo_multisample = new Gl_300_fbo_multisample();
    }

    public Gl_300_fbo_multisample() {
        super("gl-300-fbo-multisample", Profile.COMPATIBILITY, 3, 0);
    }

    private final String SHADERS_SOURCE = "image-2d";
    private final String SHADERS_ROOT = "src/data/gl_300";
    private final String TEXTURE_DIFFUSE = "kueken7_rgba8_srgb.dds";
    private Vec2i FRAMEBUFFER_SIZE = new Vec2i(160, 120);
    // With DDS textures, v texture coordinate are reversed, from top to bottom
    private int vertexCount = 6;
    private int vertexSize = vertexCount * Vec4.SIZE;
    private float[] vertexData = new float[]{
        -2.0f, -1.5f, 0.0f, 0.0f,
        +2.0f, -1.5f, 1.0f, 0.0f,
        +2.0f, +1.5f, 1.0f, 1.0f,
        +2.0f, +1.5f, 1.0f, 1.0f,
        -2.0f, +1.5f, 0.0f, 1.0f,
        -2.0f, -1.5f, 0.0f, 0.0f
    };

    private class Texture {

        private static final int DIFFUSE = 0;
        private static final int RESOLVE = 1;
        private static final int MAX = 2;
    }

    private class Framebuffer {

        private static final int RENDER = 0;
        private static final int RESOLVE = 1;
        private static final int MAX = 2;
    }

    private int programName, uniformMvp, uniformDiffuse;
    private IntBuffer vertexArrayName = GLBuffers.newDirectIntBuffer(1), bufferName = GLBuffers.newDirectIntBuffer(1),
            textureName = GLBuffers.newDirectIntBuffer(Texture.MAX),
            colorRenderbufferName = GLBuffers.newDirectIntBuffer(1),
            framebufferName = GLBuffers.newDirectIntBuffer(Framebuffer.MAX);

    @Override
    protected boolean begin(GL gl) {

        GL3 gl3 = (GL3) gl;

        boolean validated = true;

        if (validated) {
            validated = initProgram(gl3);
        }
        if (validated) {
            validated = initBuffer(gl3);
        }
        if (validated) {
            validated = initVertexArray(gl3);
        }
        if (validated) {
            validated = initTexture(gl3);
        }
        if (validated) {
            validated = initFramebuffer(gl3);
        }

        return validated & checkError(gl, "begin");
    }

    private boolean initProgram(GL3 gl3) {

        boolean validated = true;

        if (validated) {

            ShaderCode vertShader = ShaderCode.create(gl3, GL_VERTEX_SHADER,
                    this.getClass(), SHADERS_ROOT, null, SHADERS_SOURCE, "vert", null, true);
            ShaderCode fragShader = ShaderCode.create(gl3, GL_FRAGMENT_SHADER,
                    this.getClass(), SHADERS_ROOT, null, SHADERS_SOURCE, "frag", null, true);

            ShaderProgram program = new ShaderProgram();
            program.add(vertShader);
            program.add(fragShader);
            program.init(gl3);

            programName = program.program();

            gl3.glBindAttribLocation(programName, Semantic.Attr.POSITION, "position");
            gl3.glBindAttribLocation(programName, Semantic.Attr.TEXCOORD, "texCoord");

            program.link(gl3, System.out);
        }
        if (validated) {

            uniformMvp = gl3.glGetUniformLocation(programName, "mvp");
            uniformDiffuse = gl3.glGetUniformLocation(programName, "diffuse");
        }
        return validated & checkError(gl3, "initProgram");
    }

    private boolean initBuffer(GL3 gl3) {

        FloatBuffer vertexBuffer = GLBuffers.newDirectFloatBuffer(vertexData);

        gl3.glGenBuffers(1, bufferName);
        gl3.glBindBuffer(GL_ARRAY_BUFFER, bufferName.get(0));
        gl3.glBufferData(GL_ARRAY_BUFFER, vertexSize, vertexBuffer, GL_STATIC_DRAW);
        gl3.glBindBuffer(GL_ARRAY_BUFFER, 0);

        BufferUtils.destroyDirectBuffer(vertexBuffer);

        return checkError(gl3, "initBuffer");
    }

    private boolean initVertexArray(GL3 gl3) {

        gl3.glGenVertexArrays(1, vertexArrayName);
        gl3.glBindVertexArray(vertexArrayName.get(0));
        {
            gl3.glBindBuffer(GL_ARRAY_BUFFER, bufferName.get(0));
            gl3.glVertexAttribPointer(Semantic.Attr.POSITION, 2, GL_FLOAT, false, Vertex_v2fv2f.SIZE, 0);
            gl3.glVertexAttribPointer(Semantic.Attr.TEXCOORD, 2, GL_FLOAT, false, Vertex_v2fv2f.SIZE, Vec2.SIZE);
            gl3.glBindBuffer(GL_ARRAY_BUFFER, 0);

            gl3.glEnableVertexAttribArray(Semantic.Attr.POSITION);
            gl3.glEnableVertexAttribArray(Semantic.Attr.TEXCOORD);
        }
        gl3.glBindVertexArray(0);

        return checkError(gl3, "initVertexArray");
    }

    private boolean initTexture(GL3 gl3) {

        try {
            jgli.Texture texture = Load.load(TEXTURE_ROOT + "/" + TEXTURE_DIFFUSE);

            gl3.glGenTextures(Texture.MAX, textureName);
            gl3.glActiveTexture(GL_TEXTURE0);
            gl3.glBindTexture(GL_TEXTURE_2D, textureName.get(Texture.DIFFUSE));
            gl3.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0);
            gl3.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, texture.levels() - 1);
            gl3.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
            gl3.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

            jgli.Gl.Format glFormat = Gl.translate(texture.format());

            for (int level = 0; level < texture.levels(); level++) {

                gl3.glTexImage2D(GL_TEXTURE_2D, level, glFormat.internal.value,
                        texture.dimensions(level)[0], texture.dimensions(level)[1], 0,
                        glFormat.external.value, glFormat.type.value, texture.data(level));
            }
        } catch (IOException ex) {
            Logger.getLogger(Gl_300_fbo_multisample.class.getName()).log(Level.SEVERE, null, ex);
        }
        return checkError(gl3, "initTexture");
    }

    private boolean initFramebuffer(GL3 gl3) {

        gl3.glGenRenderbuffers(1, colorRenderbufferName);
        gl3.glBindRenderbuffer(GL_RENDERBUFFER, colorRenderbufferName.get(0));
        // The second parameter is the number of samples.
        gl3.glRenderbufferStorageMultisample(GL_RENDERBUFFER, 8, GL_RGBA8, FRAMEBUFFER_SIZE.x, FRAMEBUFFER_SIZE.y);

        gl3.glGenFramebuffers(Framebuffer.MAX, framebufferName);
        gl3.glBindFramebuffer(GL_FRAMEBUFFER, framebufferName.get(Framebuffer.RENDER));
        gl3.glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_RENDERBUFFER, colorRenderbufferName.get(0));
        if (!isFramebufferComplete(gl3, framebufferName.get(Framebuffer.RENDER))) {
            return false;
        }
        gl3.glBindFramebuffer(GL_FRAMEBUFFER, 0);

        gl3.glBindTexture(GL_TEXTURE_2D, textureName.get(Texture.RESOLVE));
        gl3.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
        gl3.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
        gl3.glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, FRAMEBUFFER_SIZE.x, FRAMEBUFFER_SIZE.y, 0, GL_RGBA, 
                GL_UNSIGNED_BYTE, null);

        gl3.glBindFramebuffer(GL_FRAMEBUFFER, framebufferName.get(Framebuffer.RESOLVE));
        gl3.glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, textureName.get(Texture.RESOLVE),
                0);
        if (!isFramebufferComplete(gl3, framebufferName.get(Framebuffer.RESOLVE))) {
            return false;
        }
        gl3.glBindFramebuffer(GL_FRAMEBUFFER, 0);

        return checkError(gl3, "initFramebuffer");
    }

    @Override
    public boolean render(GL gl) {

        GL3 gl3 = (GL3) gl;

        // Clear the framebuffer
        gl3.glBindFramebuffer(GL_FRAMEBUFFER, 0);
        gl3.glClearBufferfv(GL_COLOR, 0, new float[]{1.0f, 0.5f, 0.0f, 1.0f}, 0);

        gl3.glUseProgram(programName);
        gl3.glUniform1i(uniformDiffuse, 0);

        // Pass 1
        // Render the scene in a multisampled framebuffer
        gl3.glEnable(GL_MULTISAMPLE);
        renderFBO(gl3, framebufferName.get(Framebuffer.RENDER));
        gl3.glDisable(GL_MULTISAMPLE);

        // Resolved multisampling
        gl3.glBindFramebuffer(GL_READ_FRAMEBUFFER, framebufferName.get(Framebuffer.RENDER));
        gl3.glBindFramebuffer(GL_DRAW_FRAMEBUFFER, framebufferName.get(Framebuffer.RESOLVE));
        gl3.glDrawBuffer(GL_COLOR_ATTACHMENT0);
        gl3.glBlitFramebuffer(
                0, 0, FRAMEBUFFER_SIZE.x, FRAMEBUFFER_SIZE.y,
                0, 0, FRAMEBUFFER_SIZE.x, FRAMEBUFFER_SIZE.y,
                GL_COLOR_BUFFER_BIT, GL_LINEAR);
        gl3.glBindFramebuffer(GL_FRAMEBUFFER, 0);

        // Pass 2
        // Render the colorbuffer from the multisampled framebuffer
        gl3.glViewport(0, 0, windowSize.x, windowSize.y);
        renderFB(gl3, textureName.get(Texture.RESOLVE));
        return true;
    }

    private void renderFBO(GL3 gl3, int framebuffer) {

        gl3.glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
        gl3.glClearColor(0.0f, 0.5f, 1.0f, 1.0f);
        gl3.glClear(GL_COLOR_BUFFER_BIT);

        Mat4 perspective = glm.perspective_((float) Math.PI * 0.25f, (float) FRAMEBUFFER_SIZE.x / FRAMEBUFFER_SIZE.y,
                0.1f, 100.0f);
        Mat4 model = new Mat4(1.0f);
        Mat4 mvp = perspective.mul(viewMat4()).mul(model);
        gl3.glUniformMatrix4fv(uniformMvp, 1, false, mvp.toFa_(), 0);

        gl3.glViewport(0, 0, FRAMEBUFFER_SIZE.x, FRAMEBUFFER_SIZE.y);

        gl3.glActiveTexture(GL_TEXTURE0);
        gl3.glBindTexture(GL_TEXTURE_2D, textureName.get(Texture.DIFFUSE));

        gl3.glBindVertexArray(vertexArrayName.get(0));
        gl3.glDrawArrays(GL_TRIANGLES, 0, vertexCount);

        checkError(gl3, "renderFBO");
    }

    private void renderFB(GL3 gl3, int texture2dName) {

        Mat4 perspective = glm.perspective_((float) Math.PI * 0.25f, (float) windowSize.x / windowSize.y, 0.1f, 100.0f);
        Mat4 model = new Mat4(1.0f);
        Mat4 mvp = perspective.mul(viewMat4()).mul(model);
        gl3.glUniformMatrix4fv(uniformMvp, 1, false, mvp.toFa_(), 0);

        gl3.glActiveTexture(GL_TEXTURE0);
        gl3.glBindTexture(GL_TEXTURE_2D, texture2dName);

        gl3.glBindVertexArray(vertexArrayName.get(0));
        gl3.glDrawArrays(GL_TRIANGLES, 0, vertexCount);

        checkError(gl3, "renderFB");
    }

    @Override
    protected boolean end(GL gl) {

        GL3 gl3 = (GL3) gl;

        gl3.glDeleteBuffers(1, bufferName);
        gl3.glDeleteProgram(programName);
        gl3.glDeleteTextures(Texture.MAX, textureName);
        gl3.glDeleteRenderbuffers(1, colorRenderbufferName);
        gl3.glDeleteFramebuffers(Framebuffer.MAX, framebufferName);
        gl3.glDeleteVertexArrays(1, vertexArrayName);
        
        BufferUtils.destroyDirectBuffer(bufferName);
        BufferUtils.destroyDirectBuffer(textureName);
        BufferUtils.destroyDirectBuffer(colorRenderbufferName);
        BufferUtils.destroyDirectBuffer(framebufferName);
        BufferUtils.destroyDirectBuffer(vertexArrayName);

        return true;
    }
}

File: src/main/kotlin/oglSamples/tests/gl300/gl-300-fbo-multisample.kt
package oglSamples.tests.gl300

import gli_.Texture2d
import gli_.gl
import gli_.gli
import glm_.glm
import glm_.mat4x4.Mat4
import glm_.vec2.Vec2
import glm_.vec2.Vec2i
import glm_.vec4.Vec4
import gln.*
import gln.glClearColor as b
import gln.glEnable as a
import gln.cap.Caps
import gln.clear.glClearBuffer
import gln.draw.glDrawArrays
import gln.framebuffer.glBindFramebuffer
import gln.glf.glf
import gln.objects.GlProgram
import gln.renderbuffer.RenderBuffer
import gln.uniform.glUniform
import gln.vertexArray.glEnableVertexAttribArray
import gln.vertexArray.glVertexAttribPointer
import oglSamples.*
import oglSamples.framework.Framework
import oglSamples.framework.GLint
import oglSamples.framework.semantic
import org.lwjgl.opengl.GL11C
import org.lwjgl.opengl.GL13C
import org.lwjgl.opengl.GL30C
import org.lwjgl.opengl.GL30C.*

fun main() {
    gl_300_fbo_multisample()()
}

class gl_300_fbo_multisample : Framework("gl-300-fbo-multisample", Caps.Profile.COMPATIBILITY, 3, 0) {

    val SHADER_SOURCE = "gl-300/image-2d"
    val TEXTURE_DIFFUSE = "kueken7_rgba8_srgb.dds"
    val FRAMEBUFFER_SIZE = Vec2i(160, 120)

    // With DDS textures, v texture coordinate are reversed, from top to bottom
    val vertexCount = 6
    val vertexSize = vertexCount * Vertex_v2v2.size
    val vertexData = vertex_v2v2_BufferOf(
            Vertex_v2v2(Vec2(-2.0f, -1.5f), Vec2(0.0f, 0.0f)),
            Vertex_v2v2(Vec2(+2.0f, -1.5f), Vec2(1.0f, 0.0f)),
            Vertex_v2v2(Vec2(+2.0f, +1.5f), Vec2(1.0f, 1.0f)),
            Vertex_v2v2(Vec2(+2.0f, +1.5f), Vec2(1.0f, 1.0f)),
            Vertex_v2v2(Vec2(-2.0f, +1.5f), Vec2(0.0f, 1.0f)),
            Vertex_v2v2(Vec2(-2.0f, -1.5f), Vec2(0.0f, 0.0f)))

    var program = GlProgram.NULL
    val vertexArray = GlVertexArray()
    val buffer = GlArrayBuffer()
    val texture = GlTexture2d()
    val colorRenderbuffer = GlRenderBuffer()
    val colorTexture = GlTexture2d()
    val framebufferRender = GlFramebuffer()
    val framebufferResolve = GlFramebuffer()
    var uniformMVP = -1
    var uniformDiffuse = -1

    override fun begin(): Boolean {

        var validated = true

        val caps = Caps(Caps.Profile.COMPATIBILITY)

        if (validated)
            validated = initProgram()
        if (validated)
            validated = initBuffer()
        if (validated)
            validated = initVertexArray()
        if (validated)
            validated = initTexture()
        if (validated)
            validated = initFramebuffer()

        return validated && checkError("begin")
    }

    fun initProgram(): Boolean {

        var validated = true

        try {

            program = GlProgram.initFromPath("$SHADER_SOURCE.vert", "$SHADER_SOURCE.frag") {

                "Position".attrib = semantic.attr.POSITION
                "Texcoord".attrib = semantic.attr.TEXCOORD
            }
        } catch (exc: Exception) {
            validated = false
        }

        if (validated) {
            uniformMVP = program.getUniformLocation("MVP")
            uniformDiffuse = program.getUniformLocation("Diffuse")
        }

        return validated && checkError("initProgram")
    }

    fun initBuffer(): Boolean {
        buffer.gen().bind {
            data(vertexData.data)
        }
        return checkError("initBuffer")
    }

    fun initTexture(): Boolean {

        val texture = Texture2d(gli.loadDds(ClassLoader.getSystemResource(TEXTURE_DIFFUSE).toURI()))
        gl.profile = gl.Profile.GL32

        this.texture.gen().bind(0) {
            levels(0, texture.levels() - 1)
            filter(linear_mmLinear, linear)

            val format = gl.translate(texture.format, texture.swizzles)
            for (level in 0 until texture.levels())
                image(level,
                        format.internal,
                        texture[level].extent(),
                        format.external, format.type,
                        texture[level].data()!!)
        }
        return checkError("initTexture")
    }

    fun initVertexArray(): Boolean {

        vertexArray.gen().bind {
            buffer.bind {
                glVertexAttribPointer(glf.pos2_tc2)
            }
            glEnableVertexAttribArray(glf.pos2_tc2)
        }

        return checkError("initVertexArray")
    }

    fun initFramebuffer(): Boolean {

        colorRenderbuffer.gen().bind {
            // The first parameter is the number of samples.
            storageMultisample(8, GL11C.GL_RGBA8, FRAMEBUFFER_SIZE)

            if (!validate(FRAMEBUFFER_SIZE, 8, GL11C.GL_RGBA8))
                return false
        }

        framebufferRender.gen().bind {
            renderbuffer(GL30C.GL_COLOR_ATTACHMENT0, colorRenderbuffer)
            if (!complete)
                return false
        }

        colorTexture.gen().bind {
            filter(nearest, nearest)
            image(GL11C.GL_RGBA8, FRAMEBUFFER_SIZE, GL11C.GL_RGBA, GL11C.GL_UNSIGNED_BYTE)
        }
        framebufferResolve.gen().bind {
            texture(GL_COLOR_ATTACHMENT0, colorTexture)
            if (!complete)
                return false
        }

        return checkError("initFramebuffer")
    }

    fun RenderBuffer.validate(expectedSize: Vec2i, expectedSamples: GLint, expectedFormat: GLint): Boolean =
            expectedSize == size && expectedSamples == samples && expectedFormat == format

    override fun render(): Boolean {

        // Clear the framebuffer
        glBindFramebuffer(0)
        glClearBuffer(GL11C.GL_COLOR, 0, Vec4(1f, 0.5f, 0f, 1f))

        program.use()
        glUniform(uniformDiffuse, 0)

        // Pass 1
        // Render the scene in a multisampled framebuffer
        glEnable(GL13C.GL_MULTISAMPLE)
        renderFBO(framebufferRender)
        glDisable(GL_MULTISAMPLE)

        // Resolved multisampling
        framebufferRender.bindRead()
        framebufferResolve.bindDraw()
        glBlitFramebuffer(FRAMEBUFFER_SIZE)
        glBindFramebuffer()

        // Pass 2
        // Render the colorbuffer from the multisampled framebuffer
        glViewport(windowSize)
        renderFB(colorTexture)

        return true
    }

    fun renderFBO(framebuffer: GlFramebuffer)    {

        framebuffer.bind()
        glClearColor(0f, 0.5f, 1f, 1f)
        glClear(GL11C.GL_COLOR_BUFFER_BIT)

        val perspective = glm.perspective(glm.πf * 0.25f, FRAMEBUFFER_SIZE.aspect, 0.1f, 100f)
        val model = Mat4(1f)
        val mvp = perspective * view * model
        glUniform(uniformMVP, mvp)

        glViewport(FRAMEBUFFER_SIZE)

        texture.bind(0)

        vertexArray.bind()
        glDrawArrays(vertexCount)

        checkError("renderFBO")
    }

    fun renderFB(texture2D: GlTexture2d)    {

        val perspective = glm.perspective(glm.πf * 0.25f, windowSize.aspect, 0.1f, 100f)
        val model = Mat4(1f)
        val mvp = perspective * view * model
        glUniform(uniformMVP, mvp)

        texture2D.bind(0)

        vertexArray.bind()
        glDrawArrays(vertexCount)

        checkError("renderFB")
    }

    override fun end(): Boolean {

        buffer.delete()
        program.delete()
        texture.delete()
        colorTexture.delete()
        colorRenderbuffer.delete()
        framebufferRender.delete()
        framebufferResolve.delete()
        vertexArray.delete()

        return true
    }
}